File: dsa/linked_list/234_Palindrome_Linked_List.java
package linked_list;

//  * Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode fastPointer = head;
        ListNode slowPointer = head;

        while (fastPointer != null && fastPointer.next != null){
            fastPointer = fastPointer.next.next;
            slowPointer = slowPointer.next;
        }

        ListNode prevNode = null;
        ListNode curNode = slowPointer;

        while (curNode != null){
            ListNode surNode = curNode.next;
            curNode.next = prevNode;
            prevNode = curNode;
            curNode = surNode;
        }

        ListNode rightNode = prevNode;
        ListNode leftNode = head;
        while (rightNode != null){
            if (rightNode.val != leftNode.val){
                return false;
            }
            rightNode = rightNode.next;
            leftNode = leftNode.next;
        }
        return true;
    }
}

File: dsa/7+8.linked_list/234_Palindrome_Linked_List.py
import typing as t
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def isPalindrome(self, head: t.Optional[ListNode]) -> bool:
        fast_pointer: ListNode = head
        slow_pointer: ListNode = head

        while fast_pointer and fast_pointer.next:
            slow_pointer = slow_pointer.next
            fast_pointer = fast_pointer.next.next
        
        prev_pointer: ListNode = None
        while slow_pointer:
            next_pointer: ListNode = slow_pointer.next
            slow_pointer.next = prev_pointer
            prev_pointer = slow_pointer
            slow_pointer = next_pointer
        
        cur_pointer: ListNode = head
        
        while prev_pointer and cur_pointer:
            if prev_pointer.val != cur_pointer.val:
                return False
            prev_pointer = prev_pointer.next
            cur_pointer = cur_pointer.next
        
        return True
        
